entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -543,9 +555,72 @@ public void total(
     responseObserver.onCompleted();
   }
 
+  @Override
+  public void deleteEntities(
+      DeleteEntitiesRequest request, StreamObserver<DeleteEntitiesResponse> responseObserver) {
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    try {
+      validateDeleteEntitiesRequest(request, requestContext);
+    } catch (Exception ex) {
+      responseObserver.onError(ex);
+      return;
+    }
+
+    List<String> entityIds = getEntityIdsToDelete(requestContext, request);
+    if (entityIds.size() == 0) {
+      LOG.debug("{}. No entities found to delete", request);
+      responseObserver.onNext(DeleteEntitiesResponse.newBuilder().build());
+      responseObserver.onCompleted();
+      return;
+    }
+
+    if (entityIds.size() > maxEntitiesToDelete) {
+      LOG.warn(
+          "{}. Number of ids to delete exceeds the maximum limit: Entity Ids limit exceeded",
+          request);
+      responseObserver.onError(new RuntimeException("Entity Ids limit exceeded"));
+      return;
+    }
+
+    LOG.debug("Deleting entity of type: {} and ids: {}", request.getEntityType(), entityIds);
+    try {
+      Optional<String> tenantId = requestContext.getTenantId();
+      this.entitiesCollection.delete(
+          DocStoreConverter.transform(tenantId.orElseThrow(), request.getEntityType(), entityIds));
+      responseObserver.onNext(
+          DeleteEntitiesResponse.newBuilder().addAllEntityIds(entityIds).build());
+      responseObserver.onCompleted();
+    } catch (Exception ex) {
+      responseObserver.onError(ex);
+    }
+  }
+
+  private List<String> getEntityIdsToDelete(
+      RequestContext requestContext, DeleteEntitiesRequest request) {
+    List<String> docStoreSelections = List.of(EntityServiceConstants.ENTITY_ID);
+    EntityQueryRequest entityQueryRequest =
+        EntityQueryRequest.newBuilder()
+            .setEntityType(request.getEntityType())
+            .setFilter(request.getFilter())
```

Comment:
For optimisation purposes, we could add a limit `maxEntitiesToDelete + 1` since we know we are never going to consume more than `maxEntitiesToDelete`.